crates/cargo-test-macro/src/lib.rs

```diff
@@ -29,57 +39,81 @@ pub fn cargo_test(attr: TokenStream, item: TokenStream) -> TokenStream {
                 // explicit opt-in (these generally only work on linux, and
                 // have some extra requirements, and are slow, and can pollute
                 // the environment since it downloads dependencies).
-                ignore |= is_not_nightly;
-                ignore |= option_env!("CARGO_RUN_BUILD_STD_TESTS").is_none();
+                set_ignore!(is_not_nightly, "requires nightly");
+                set_ignore!(
+                    option_env!("CARGO_RUN_BUILD_STD_TESTS").is_none(),
+                    "CARGO_RUN_BUILD_STD_TESTS must be set"
+                );
             }
             "build_std_mock" => {
                 // Only run the "mock" build-std tests on nightly and disable
                 // for windows-gnu which is missing object files (see
                 // https://github.com/rust-lang/wg-cargo-std-aware/issues/46).
-                ignore |= is_not_nightly;
-                ignore |= cfg!(all(target_os = "windows", target_env = "gnu"));
+                set_ignore!(is_not_nightly, "requires nightly");
+                set_ignore!(
+                    cfg!(all(target_os = "windows", target_env = "gnu")),
+                    "does not work on windows-gnu"
+                );
             }
             "nightly" => {
                 requires_reason = true;
-                ignore |= is_not_nightly;
+                set_ignore!(is_not_nightly, "requires nightly");
             }
             s if s.starts_with("requires_") => {
                 let command = &s[9..];
-                ignore |= !has_command(command);
+                set_ignore!(!has_command(command), "{command} not installed");
             }
             s if s.starts_with(">=1.") => {
                 requires_reason = true;
                 let min_minor = s[4..].parse().unwrap();
-                ignore |= version().0 < min_minor;
+                let minor = version().0;
+                set_ignore!(minor < min_minor, "requires rustc 1.{minor} or newer");
             }
             s if s.starts_with("reason=") => {
-                found_reason = true;
+                explicit_reason = Some(s[7..].parse().unwrap());
             }
             _ => panic!("unknown rule {:?}", rule),
         }
     }
-    if requires_reason && !found_reason {
+    if requires_reason && explicit_reason.is_none() {
         panic!(
             "#[cargo_test] with a rule also requires a reason, \
             such as #[cargo_test(nightly, reason = \"needs -Z unstable-thing\")]"
         );
     }
```

Comment:
Would it make sense to add `ignore` to `#[cargo_test]` so that it would require a reason?

```rust
#[cargo_test(ignore, reason = "broken, need artifact info in index")]
```

Reply:
It could, but it wouldn't prevent someone from adding `#[ignore]` and would also make it tricky to write conditional ignores, like:

```rust
#[cfg_attr(
    any(target_env = "msvc", target_vendor = "apple"),
    ignore = "--out-dir and examples are currently broken on MSVC and apple"
)]
```

I wouldn't want to complicate the macro too much, or to have multiple ways of ignoring (`#[cargo_test(ignore)]` and `#[ignore]`).

For the most part, these ignore reasons are just a convenience. I don't see it critical that they get printed 100% of the time.

I think in my ideal world, tests could be ignored from within the test using code instead of using attributes, which are not well suited for complex conditions.